flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java

```diff
@@ -0,0 +1,398 @@
+				// 1. After processing a record with offset of "stoppingOffset - 1". The split reader
+				//    should not continue fetching because the record with stoppingOffset may not exist.
+				// 2. Before processing a record whose offset is greater than or equals to the stopping
+				//    offset. This should only happens when case 1 was not met due to log compaction or
+				//    log retention.
+				// Case 2 is handled here. Case 1 is handled after the record is processed.
+				if (consumerRecord.offset() >= stoppingOffset) {
+					finishSplitAtRecord(tp, stoppingOffset, consumerRecord.offset(),
+							finishedPartitions, recordsBySplits);
+					break;
+				}
+				// Add the record to the partition collector.
+				try {
+					deserializationSchema.deserialize(consumerRecord, collector);
+					collector.getRecords().forEach(r ->
+							recordsForSplit.add(new Tuple3<>(r,
+									consumerRecord.offset(),
+									consumerRecord.timestamp())));
+					// Finish the split because there might not be any message after this point. Keep polling
+					// will just block forever.
+					if (consumerRecord.offset() == stoppingOffset - 1) {
+						finishSplitAtRecord(tp, stoppingOffset, consumerRecord.offset(),
+								finishedPartitions, recordsBySplits);
+					}
+				} catch (Exception e) {
+					throw new IOException("Failed to deserialize consumer record due to", e);
+				} finally {
+					collector.reset();
+				}
+			}
+		}
+		// Unassign the partitions that has finished.
+		if (!finishedPartitions.isEmpty()) {
+			unassignPartitions(finishedPartitions);
+		}
+		recordsBySplits.prepareForRead();
+		return recordsBySplits;
+	}
+
+	@Override
+	public void handleSplitsChanges(SplitsChange<KafkaPartitionSplit> splitsChange) {
+		// Get all the partition assignments and stopping offsets.
+		if (!(splitsChange instanceof SplitsAddition)) {
+			throw new UnsupportedOperationException(String.format(
+				"The SplitChange type of %s is not supported.", splitsChange.getClass()));
+		}
+
+		// Assignment.
+		List<TopicPartition> newPartitionAssignments = new ArrayList<>();
+		// Starting offsets.
+		Map<TopicPartition, Long> partitionsStartingFromSpecifiedOffsets = new HashMap<>();
+		List<TopicPartition> partitionsStartingFromEarliest = new ArrayList<>();
+		List<TopicPartition> partitionsStartingFromLatest = new ArrayList<>();
+		// Stopping offsets.
+		List<TopicPartition> partitionsStoppingAtLatest = new ArrayList<>();
+		Set<TopicPartition> partitionsStoppingAtCommitted = new HashSet<>();
+
+		// Parse the starting and stopping offsets.
+		splitsChange.splits().forEach(s -> {
+			newPartitionAssignments.add(s.getTopicPartition());
+			parseStartingOffsets(s, partitionsStartingFromEarliest, partitionsStartingFromLatest, partitionsStartingFromSpecifiedOffsets);
+			parseStoppingOffsets(s, partitionsStoppingAtLatest, partitionsStoppingAtCommitted);
+		});
+
+		// Assign new partitions.
+		newPartitionAssignments.addAll(consumer.assignment());
```

Comment:
I thought `splitsChange` always contains the complete assignment (not delta), as KafkaSubscriber impl puts current assignment as removed partitions. if my understanding is correct, this newPartitionAssignments list can grow forever for each partition discovery cycle.

Reply:
thx for explaining. I missed it earlier.